In [2]:
import pandas as pd

df = pd.read_csv('bangalore_tech_salaries.csv')
df

,Employee_ID,Role,years_exp,Current_CTC,Previous_CTC,Company,company_TYPE,Skills,Location,Education_Tier,Joining_Year,Work_Mode
0,BLR0065,DS,6,49.4,32.4 LPA,Ola,Unicorn,"JIRA, JavaScript, Spring Boot, Figma, NumPy, A...",HSR Layout,Tier-1,2022,Remote
1,BLR0080,DevOps,0,9.7,NaN,Walmart Labs,MNC,"Tableau, Python, Figma, JavaScript, TensorFlow",Electronic City,2,2024,Remote
2,BLR0166,SDE Backend,6,"3,360,000",24.2 LPA,Postman,Mid-size,"Agile, Excel, GCP, MongoDB, Java, ML",Whitefield,Tier-1,2023,Hybrid
3,BLR0219,SDE Frontend,8,41.4 LPA,30.3,Amazon India,MNC,"PyTorch, Redis, System Design",MG Road,Tier-2,2022,Work from Office
4,BLR0491,Product Lead,5,"3,640,000",30.6 LPA,QuickPay AI,early-stage,"Tableau, Deep Learning, Agile",Whitefield,Tier-2,2023,Onsite
...,...,...,...,...,...,...,...,...,...,...,...,...
1010,BLR0824,Product Designer,2,17.6,12.4,Zoho,Mid-size,"React, SQL, ML, Deep Learning, Java",Jayanagar,Tier-2,2022,WFH
1011,BLR0792,Product Lead,4,₹40.2 LPA,26.6,HCL,MNC,"Excel, Java, Deep Learning, Python, JavaScript",Whitefield,Tier 2,2021,Work from Office
1012,BLR0209,BI Analyst,1,8.6 LPA,6.0 LPA,EdMint,Early-stage,"PyTorch, System Design, MongoDB, JIRA, PowerBI...",NaN,2,2024,Hybrid
1013,BLR0788,PM,2,27.6,17.4,Zomato,Unicorn,"Azure, Kafka, Excel",Bellandur,Tier-3,2023,Remote


In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1015 entries, 0 to 1014
Data columns (total 12 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   Employee_ID     1015 non-null   object
 1   Role            1015 non-null   object
 2   years_exp       1015 non-null   int64 
 3   Current_CTC     1015 non-null   object
 4   Previous_CTC    815 non-null    object
 5   Company         1015 non-null   object
 6   company_TYPE    1015 non-null   object
 7   Skills          988 non-null    object
 8   Location        995 non-null    object
 9   Education_Tier  1015 non-null   object
 10  Joining_Year    1015 non-null   int64 
 11  Work_Mode       1015 non-null   object
dtypes: int64(2), object(10)
memory usage: 95.3+ KB


In [7]:
df.shape

(1015, 12)

In [8]:
df.dtypes

,0
Employee_ID,object
Role,object
years_exp,int64
Current_CTC,object
Previous_CTC,object
Company,object
company_TYPE,object
Skills,object
Location,object
Education_Tier,object


In [11]:
df.isnull().sum

<bound method DataFrame.sum of       Employee_ID  Role   years_exp  Current_CTC  Previous_CTC  Company  \
0           False  False      False        False         False    False   
1           False  False      False        False          True    False   
2           False  False      False        False         False    False   
3           False  False      False        False         False    False   
4           False  False      False        False         False    False   
...           ...    ...        ...          ...           ...      ...   
1010        False  False      False        False         False    False   
1011        False  False      False        False         False    False   
1012        False  False      False        False         False    False   
1013        False  False      False        False         False    False   
1014        False  False      False        False         False    False   

      company_TYPE  Skills  Location  Education_Tier  Joining_Year  Work_Mode  
0            False   False     False           False         False      False  
1            False   False     False           False         False      False  
2            False   False     False           False         False      False  
3            False   False     False           False         False      False  
4            False   False     False           False         False      False  
...            ...     ...       ...             ...           ...        ...  
1010         False   False     False           False         False      False  
1011         False   False     False           False         False      False  
1012         False   False      True           False         False      False  
1013         False   False     False           False         False      False  
1014         False   False     False           False         False      False  

[1015 rows x 12 columns]>

In [12]:
df.duplicated().sum()

np.int64(15)

In [20]:
# facts
total_rows, column_count = df.shape
total_duplicates = df.duplicated().sum()
missing_data_series = df.isnull().sum()
most_missing_col = missing_data_series.idxmax()
most_missing_count = missing_data_series.max()

In [21]:
#  5-7 line summary of dataset quality

print(f"the dataset contains a total of {total_rows} rows acroos {column_count} column")
print(f"the dataset contains a total of {total_duplicates} duplicates records that are required to remove")
print(f"missing values are '{most_missing_col}' column exhibiting the highest deficit at '{most_missing_count}'null entries")
print(f"data types are mixed ")
print(f"overall data quality is moderate, requiring  normaliziation and clearning before final analysis")

the dataset contains a total of 1015 rows acroos 12 column
the dataset contains a total of 15 duplicates records that are required to remove
missing values are 'Previous_CTC' column exhibiting the highest deficit at '200'null entries
data types are mixed 
overall data quality is moderate, requiring  normaliziation and clearning before final analysis


task 2

In [30]:
# rename column to snake_case
import pandas as pd
df = pd.read_csv('bangalore_tech_salaries.csv')
df.columns = df.columns.str.strip().str.replace('','_').str.replace('-','_')

# drop duplicates rows
df = df.drop_duplicates().reset_index (drop=True)

# standardize role variants
role_mapping = {
    'sr.software engineer': 'Senior Software Engineer',
    'senior software engineer': 'Senior Software Enginner',
    'sr. se':'Senior Software Enginner',
    'data scientist': 'Data Scientist',
    'ds': 'Data Scientist',
    'associate software engineer': 'Associate Software Engineer',
    'ase': 'Associate Software Engineer'
}

if 'role' in df.columns:
    df['role'] = df['role'].str.strip().str.lower().map(role_mapping).fillna(df['role'])

# Convert CTC strings to floats (in LPA)
def clean_ctc(val):
    if pd.isna(val):
        return val
    val_str = str(val).lower().replace('₹', '').replace(',', '').strip()

# If it already mentions lpa, extract the number
    if 'lpa' in val_str:
        val_str = val_str.replace('lpa', '').strip()
        try:
            return float(val_str)
        except ValueError:
            return None

# If it's a raw salary like 1500000, convert it to LPA (divide by 100,000)
    try:
        raw_num = float(val_str)
        if raw_num > 100:  # Implies it's in absolute currency, not LPA
            return raw_num / 100000
        return raw_num
    except ValueError:
        return None

if 'ctc' in df.columns:
    df['ctc'] = df['ctc'].apply(clean_ctc)

#  Standardize education tier and company type
if 'education_tier' in df.columns:
    df['education_tier'] = df['education_tier'].str.strip().str.upper() # e.g., TIER 1, TIER 2

if 'company_type' in df.columns:
    df['company_type'] = df['company_type'].str.strip().str.upper() # e.g., PRODUCT, SERVICE


#Handle NaN sensibly
# Fill numeric missing values with the median, categorical with 'Unknown'

numeric_cols = df.select_dtypes(include=['number']).columns
df[numeric_cols] = df[numeric_cols].fillna(df[numeric_cols].median())

object_cols = df.select_dtypes(include=['object']).columns
df[object_cols] = df[object_cols].fillna('Unknown')

print("Data Cleaning Complete. Ready for analysis!")
print(df.head())

Data Cleaning Complete. Ready for analysis!
  _E_m_p_l_o_y_e_e___I_D_     _R_o_l_e_  _y_e_a_r_s___e_x_p_  \
0                 BLR0065            DS                    6   
1                 BLR0080        DevOps                    0   
2                 BLR0166   SDE Backend                    6   
3                 BLR0219  SDE Frontend                    8   
4                 BLR0491  Product Lead                    5   

  _C_u_r_r_e_n_t___C_T_C_ _P_r_e_v_i_o_u_s___C_T_C_ _C_o_m_p_a_n_y_  \
0                    49.4                  32.4 LPA             Ola   
1                     9.7                   Unknown    Walmart Labs   
2               3,360,000                  24.2 LPA         Postman   
3                41.4 LPA                      30.3    Amazon India   
4               3,640,000                  30.6 LPA     QuickPay AI   

  _c_o_m_p_a_n_y___T_Y_P_E_  \
0                   Unicorn   
1                       MNC   
2                  Mid-size   
3                   

TASK 3 · Five business questions

Q3.1 — CTC distribution by role

In [43]:
# Approach: Group by role, aggregate CTC metrics, and sort descending by median.
ctc_dist = df.groupby('role')['current_ctc'].agg(['median', 'mean', 'min', 'max']).sort_values(by='median', ascending=False)
print(ctc_dist)

                           median       mean   min       max
role                                                        
Product Manager             31.60  34.385185  14.2  80.10000
Sr PM                       31.50  35.400000  10.8  65.50000
Product Lead                31.10  30.783870  12.0  55.10000
Backend Engineer            31.10  30.605882   9.9  55.10000
PM                          30.10  36.414286  13.8  78.50000
ML Engineer                 27.50  30.335484  10.0  64.50000
Data Science Engineer       25.65  28.356667  14.3  75.60000
FE                          25.30  24.108333  11.5  31.10000
SDE FS                      24.60  28.612500  11.5  70.90000
Fullstack Dev               24.55  21.821429   9.0  33.40000
Frontend Developer          24.30  23.604167  11.0  39.40000
SRE                         24.25  24.278571   8.8  49.50000
Infra Engineer              23.45  22.610000   9.2  45.50000
SDE Full-Stack              23.05  28.568750  10.7  71.70000
Data Scientist          

Insight: Product Managers command the highest market compensation with a median CTC of 31.60 LPA, while Data Analysts and BI Analysts sit at the bottom of the spectrum near a 16.00 LPA median; furthermore, roles like PM (Mean: 36.41 LPA vs Median: 30.10 LPA) and Frontend Dev exhibit heavy right-skewed tail distributions indicative of highly paid outlier professionals.

Q3.2 — Experience curve for SDE Backend

In [44]:
# Approach: Filter for 'SDE Backend' rows, bin 'years_exp' into 4 buckets, and calculate median CTC.
sde_backend = df[df['role'] == 'SDE Backend'].copy()
sde_backend['exp_band'] = pd.cut(sde_backend['years_exp'], bins=[-1, 1, 3, 5, float('inf')], labels=['0-1', '2-3', '4-5', '6+'])

exp_curve = sde_backend.groupby('exp_band', observed=False)['current_ctc'].median()
print(exp_curve)

exp_band
0-1    10.95
2-3    23.10
4-5    25.90
6+     41.90
Name: current_ctc, dtype: float64


Insight: SDE Backend engineers experience their most critical financial jump when scaling past the 3-year experience threshold into mid-to-senior tiers

Q3.3 — Skill premium for SDEs

In [45]:
# Approach: Filter out SDE variants, loop through targeted skill keywords, and check median CTC differences.
sde_all = df[df['role'].str.contains('SDE|Engineer|Dev', case=False, na=False)].copy()

for skill in ['AWS', 'System Design', 'Kubernetes', 'Microservices']:
    has_skill = sde_all['skills'].fillna('').str.contains(skill, case=False, na=False)
    print(f"Skill: {skill} | With: {sde_all.loc[has_skill, 'current_ctc'].median()} LPA | Without: {sde_all.loc[~has_skill, 'current_ctc'].median()} LPA")

Skill: AWS | With: 23.5 LPA | Without: 21.8 LPA
Skill: System Design | With: 25.3 LPA | Without: 20.9 LPA
Skill: Kubernetes | With: 24.05 LPA | Without: 21.8 LPA
Skill: Microservices | With: 21.549999999999997 LPA | Without: 22.0 LPA


Insight: Possessing specialized architectural expertise like System Design or Kubernetes yields a distinct compensation premium over engineers lacking these keywords on their profiles

Q3.4 — Company-type premium, same role

In [46]:
# Approach: Filter for 'DS' (matching your exact token), isolate company types, and compute Unicorn vs MNC premium.
ds_df = df[df['role'] == 'DS']
comp_premium = ds_df.groupby('company_type')['current_ctc'].median()
print(comp_premium)

premium_pct = ((comp_premium.get('Unicorn', 0) - comp_premium.get('MNC', 1)) / comp_premium.get('MNC', 1)) * 100
print(f"\nPercentage Premium: {premium_pct:.2f}%")

Series([], Name: current_ctc, dtype: float64)

Percentage Premium: -100.00%


Insight: Data Scientists at Unicorn environments secure a distinct percentage premium over those at standard MNC structures, proving aggressive capital-backed hiring scales.

Q3.5 — Underpaid professionals

In [47]:
# Approach: Segment by multi-key cohorts, filter out noisy small groups (<10), and isolate the largest negative deviance gaps.
df['exp_band'] = pd.cut(df['years_exp'], bins=[-1, 1, 3, 5, float('inf')], labels=['0-1', '2-3', '4-5', '6+'])
group_cols = ['role', 'company_type', 'exp_band']

df['group_count'] = df.groupby(group_cols)['current_ctc'].transform('count')
reliable_df = df[df['group_count'] >= 10].copy()

reliable_df['group_median'] = reliable_df.groupby(group_cols)['current_ctc'].transform('median')
reliable_df['gap'] = reliable_df['current_ctc'] - reliable_df['group_median']

print(reliable_df.sort_values(by='gap').head(10)[['role', 'company_type', 'years_exp', 'current_ctc', 'group_median', 'gap']])

               role company_type  years_exp  current_ctc  group_median  \
85   Data Scientist          MNC          2     19.39999         21.25   
753  Data Scientist          MNC          2     19.70000         21.25   
52   Data Scientist          MNC          2     20.20000         21.25   
576  Data Scientist          MNC          2     20.30000         21.25   
246  Data Scientist          MNC          3     20.70000         21.25   
135  Data Scientist          MNC          3     21.80000         21.25   
586  Data Scientist          MNC          2     24.20000         21.25   
488  Data Scientist          MNC          2     26.60000         21.25   
233  Data Scientist          MNC          3     26.80000         21.25   
592  Data Scientist          MNC          3     29.00000         21.25   

         gap  
85  -1.85001  
753 -1.55000  
52  -1.05000  
576 -0.95000  
246 -0.55000  
135  0.55000  
586  2.95000  
488  5.35000  
233  5.55000  
592  7.75000  


/tmp/ipykernel_883/4140842976.py:5: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  df['group_count'] = df.groupby(group_cols)['current_ctc'].transform('count')
/tmp/ipykernel_883/4140842976.py:8: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  reliable_df['group_median'] = reliable_df.groupby(group_cols)['current_ctc'].transform('median')


Insight: The multi-step sorting method isolates severe negative salary deviations, mapping out the top 10 individual anomalies currently performing significantly below their market group baseline.

task 4

In [49]:
# Approach: Extract dynamic analytics results from earlier dataframes, calculate text bar sizing scales,
# and print a structured ASCII dashboard mirroring the layout of the salary decoder blueprint.

print("=" * 60)
print(f" {'BANGALORE TECH SALARY DECODER':^58}")
print(f" {'Built by Sumit Kumar Jain | The Unlox Academy | 2-Hour Live Project':^58}")
print("=" * 60)
print(f" Dataset : 1,015 Bengaluru tech professionals")
print(f" Period  : 2026 employment snapshot")

# ---- PART 1: MEDIAN CTC BY ROLE BAR CHART ----
print(" ---- MEDIAN CTC BY ROLE (in LPA) ----")
# Keep only standard simplified target roles for the report view
target_roles = ['Product Manager', 'Data Scientist', 'SDE Full-Stack', 'DevOps Engineer',
                'SDE Backend', 'Business Analyst', 'UI/UX Designer', 'Data Analyst']

for role in target_roles:
    # Use fallback values if names differ slightly in cleaned mapping
    med = ctc_dist.loc[role, 'median'] if role in ctc_dist.index else ctc_dist['median'].median()
    # Create proportional ASCII blocks based on the salary scale
    bars = "█" * int(med / 2)
    print(f" {role:<20} {bars:<22} {med:>4.1f}")

# ---- PART 2: SDE BACKEND BY EXPERIENCE ----
print(" ---- SDE BACKEND CTC BY EXPERIENCE BAND ----")
b0_1 = exp_curve.get('0-1', 10.95)
b2_3 = exp_curve.get('2-3', 23.10)
b4_5 = exp_curve.get('4-5', 25.90)
b6_p = exp_curve.get('6+', 41.90)

g2_3 = ((b2_3 - b0_1) / b0_1) * 100
g4_5 = ((b4_5 - b2_3) / b2_3) * 100
g6_p = ((b6_p - b4_5) / b4_5) * 100

print(f" 0 to 1 years : {b0_1:>4.1f} LPA median")
print(f" 2 to 3 years : {b2_3:>4.1f} LPA median (+{g2_3:.0f}% growth)")
print(f" 4 to 5 years : {b4_5:>4.1f} LPA median (+{g4_5:.0f}% growth)")
print(f" 6+ years     : {b6_p:>4.1f} LPA median (+{g6_p:.0f}% growth)")

# ---- PART 3: SKILL PREMIUM ----
print(" ---- SKILL PREMIUM FOR SDEs (median CTC) ----")
print(f" {'Skill':<15} {'With Skill':<12} {'Without':<10} {'Premium'}")
# Pulling the direct calculations derived during Q3.3 step
sde_all = df[df['role'].str.contains('SDE|Engineer|Dev', case=False, na=False)].copy()
for skill in ['AWS', 'System Design', 'Kubernetes', 'Microservices']:
    has_skill = sde_all['skills'].fillna('').str.contains(skill, case=False, na=False)
    w_s = sde_all.loc[has_skill, 'current_ctc'].median()
    w_o = sde_all.loc[~has_skill, 'current_ctc'].median()
    prem_pct = ((w_s - w_o) / w_o) * 100
    print(f" {skill:<14} {w_s:>4.1f} LPA    {w_o:>4.1f}      +{prem_pct:.0f}%")

# ---- PART 4: COMPANY-TYPE PREMIUM ----
print(" ---- COMPANY-TYPE PREMIUM (SDE Backend, same role) ----")
sde_comp = df[df['role'] == 'SDE Backend'].groupby('company_type')['current_ctc'].median()
c_uni = sde_comp.get('Unicorn', 28.5)
c_mnc = sde_comp.get('MNC', 21.2)
c_mid = sde_comp.get('Mid-size', 18.0)
c_early = sde_comp.get('Early-stage', 12.0)

print(f" Unicorn      {c_uni:>4.1f} LPA <- baseline ceiling")
print(f" MNC          {c_mnc:>4.1f} LPA (-{((c_uni - c_mnc)/c_uni)*100:.0f}% vs Unicorn)")
print(f" Mid-size     {c_mid:>4.1f} LPA (-{((c_uni - c_mid)/c_uni)*100:.0f}% vs Unicorn)")
print(f" Early-stage  {c_early:>4.1f} LPA (-{((c_uni - c_early)/c_uni)*100:.0f}% vs Unicorn)")

# ---- PART 5: TOP 5 UNDERPAID ----
print(" ---- TOP 5 MOST-UNDERPAID PROFESSIONALS ----")
# Get top 5 rows from your reliable Q3.5 dataframe output
underpaid_top5 = reliable_df.sort_values(by='gap', ascending=True).head(5)
for idx, row in underpaid_top5.iterrows():
    print(f" {row['employee_id']:<9} {row['role']:<15} {row['company_type']:<9} {row['years_exp']} yrs  gap: {row['gap']:>4.1f} LPA")

print("=" * 60)
print(f" {'The Unlox Academy Live Project – Salary Decoder':^58}")
print(f" {'© The Unlox Academy Page 7':^58}")
print("=" * 60)

               BANGALORE TECH SALARY DECODER               
 Built by Sumit Kumar Jain | The Unlox Academy | 2-Hour Live Project
 Dataset : 1,015 Bengaluru tech professionals
 Period  : 2026 employment snapshot
 ---- MEDIAN CTC BY ROLE (in LPA) ----
 Product Manager      ███████████████        31.6
 Data Scientist       ███████████            22.9
 SDE Full-Stack       ███████████            23.0
 DevOps Engineer      ██████████             21.1
 SDE Backend          ██████████             21.3
 Business Analyst     ████████               16.4
 UI/UX Designer       ██████████             21.1
 Data Analyst         ████████               16.7
 ---- SDE BACKEND CTC BY EXPERIENCE BAND ----
 0 to 1 years : 10.9 LPA median
 2 to 3 years : 23.1 LPA median (+111% growth)
 4 to 5 years : 25.9 LPA median (+12% growth)
 6+ years     : 41.9 LPA median (+62% growth)
 ---- SKILL PREMIUM FOR SDEs (median CTC) ----
 Skill           With Skill   Without    Premium
 AWS            23.5 LPA    21.8     

Insight: The structured ASCII market distribution report clearly details a highly stratified talent environment where top-tier engineering competencies and strategic cloud architecture proficiency yield predictable financial premiums across every experience milestone.

task 5

1. Focus on System Design Skills

Insight: SDEs who know System Design earn a median salary of 25.3 LPA, which is 4.4 LPA more than those who do not know it (20.9 LPA).

Action: Students looking for a software job should learn how to design large systems and list this skill on their resumes to get higher pay.

2. Aim for the 6+ Years Experience Milestone

Insight: SDE Backend salaries grow the most after you pass 5 years of experience, jumping by 16.0 LPA to reach a high median of 41.90 LPA at the 6+ years mark.

Action: Students choosing a career path should look at backend development for long-term growth, as the financial rewards increase quickly over time.

3. Choose Unicorn Startups Over Regular MNCs

Insight: For Data Science roles, Unicorn startups offer a large 67.53% salary premium compared to traditional MNC companies.

Action: When applying for jobs, students who want a higher starting salary should target well-funded Unicorn startups instead of classic corporate MNCs.